### PR TITLE
Incremental 02a - Unit test for kokkos_malloc 

### PR DESCRIPTION
**Pull request by rgayatri23**

### CODE REVIEW

**Comment by dalg24** [comment]
Add to whitelist

File: core/unit_test/Makefile [hunk 1]
```diff
@@ -306,6 +306,8 @@ endif
 
 ifeq ($(KOKKOS_INTERNAL_USE_SERIAL), 1)
     OBJ_SERIAL = UnitTestMainInit.o gtest-all.o
+    OBJ_SERIAL += TestSerial_Incremental_01_ExecutionSpace.o
```
Comment (jeffmiles63) [inline core/unit_test/Makefile]: I don't think that incremental tests need to be added to the existing targets in the GNU Makefile

File: core/unit_test/serial/TestSerial_Incremental_01_ExecutionSpace.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
+
```
Comment (jeffmiles63) [inline core/unit_test/serial/TestSerial_Incremental_01_ExecutionSpace.cpp]: cpp file is not necessary...we will generate these

File: core/unit_test/serial/TestSerial_Incremental_02a_MemorySpace.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
+
```
Comment (jeffmiles63) [inline core/unit_test/serial/TestSerial_Incremental_02a_MemorySpace.cpp]: cpp file is not necessary...we will generate these

File: core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+template <class ExecSpace>
+struct TestIncrExecSpaceTypedef {
+
+  void testit() {
+    bool passed = true;
+    passed &= !std::is_same<void,typename ExecSpace::memory_space>::value;
+    passed &=  std::is_same<ExecSpace,typename ExecSpace::execution_space>::value;
+    passed &= !std::is_same<void,typename ExecSpace::scratch_memory_space>::value;
+    passed &= !std::is_same<void,typename ExecSpace::array_layout>::value;
+    ASSERT_TRUE(passed);
```
Comment (dalg24) [inline core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp]: Split into multiple assertions for readability.

File: core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+namespace Test {
+
+// Test construction and assignment
+
+template <class ExecSpace>
+struct TestIncrExecSpaceTypedef {
+
+  void testit() {
+    bool passed = true;
+    passed &= !std::is_same<void,typename ExecSpace::memory_space>::value;
```
Comment (jeffmiles63) [inline core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp]: As a matter of process, all of the source files included in pull requests into kokkos/develop must be formatted using the clang format tool.  The way to accomplish this is to run the ./scripts/apply-clang-format script from the kokkos directory.  The format expected is what is generated using clang 8.0.

File: core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <cstdio>
+#include <sstream>
+
+namespace Test {
+
+// Test construction and assignment
```
Comment (dalg24) [inline core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp]: There is no construction nor assignment in these tests.

File: core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <cstdio>
+#include <sstream>
```
Comment (dalg24) [inline core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp]: You are not using any of these two standard headers.
You need to include `<type_traits>` and `<gtest/gtest.h>`

File: core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    typedef typename device_type::memory_space memory_space;
+    typedef typename device_type::execution_space execution_space;
+
+    passed &= std::is_same<device_type, Kokkos::Device<execution_space, memory_space>>::value;
+
+    ASSERT_TRUE(passed);
+  }
+};
+
+TEST_F(TEST_CATEGORY, incr_01_execspace_typedef) {
```
Comment (dalg24) [inline core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp]: Do not use fixtures if you don't use them.  Use the `TEST()` macro instead.

File: core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  TestIncrExecSpaceTypedef<TEST_EXECSPACE> test;
+  test.testit();
+}
+
+TEST_F(TEST_CATEGORY, incr_01_execspace) {
+  ASSERT_TRUE(Kokkos::is_execution_space<TEST_EXECSPACE>::value);
+  ASSERT_FALSE(Kokkos::is_execution_space<TestIncrExecSpaceTypedef<TEST_EXECSPACE>>::value);
```
Comment (dalg24) [inline core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp]: What is the point of that assertion?

File: core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    passed &= !std::is_same<void,typename ExecSpace::scratch_memory_space>::value;
+    passed &= !std::is_same<void,typename ExecSpace::array_layout>::value;
+    ASSERT_TRUE(passed);
+  }
+};
+
+template <class ExecSpace>
+struct TestIncrExecSpace {
+
+  void testit() {
```
Comment (dalg24) [inline core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp]: Is not called anywhere.

**Comment by jeffmiles63** [comment]
We are actually planning to automate the cmake and gnu make logic for adding the tests.  These are the tests that will be referenced in #2533 (still WIP of course)

**Comment by jeffmiles63** [comment]
> Unit tests need to be added to cmake.
We are actually planning to automate the cmake and gnu make logic for adding the tests. These are the tests that will be referenced in #2533 (still WIP of course)

File: core/unit_test/incremental_testing/TestIncremental_02_MemorySpace.hpp [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+  void testit_free() {
+
+    const int N = 100000;
+    const int M = 100000;
+    for(int i = 0; i < N; ++i)
+    {
+      double *data = (double*) Kokkos::kokkos_malloc<MemSpace>("data", M*sizeof(double));
+      ASSERT_FALSE(data == nullptr);
+      Kokkos::kokkos_free<MemSpace>(data);
+    }
```
Comment (dalg24) [inline core/unit_test/incremental_testing/TestIncremental_02_MemorySpace.hpp]: Please add a comment about the intent here.  I assume you allocate many times enough memory such that the device would run out of memory if de-allocation does not happen.  The cumulated amount of memory allocated is 80GB (assuming double are 8 bytes).

File: core/unit_test/incremental_testing/TestIncremental_02_MemorySpace.hpp [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+struct TestIncrMemorySpace_deepcopy {
+
+  const int num_elements = 10;
+  typedef Kokkos::View<double*, Kokkos::LayoutRight, MemSpaceA> viewDevice;
+  typedef Kokkos::View<double*, Kokkos::LayoutLeft, MemSpaceB> viewHost;
+
+  //fill the host side with 0.5 and device side with 0.0
+  void fill_host(viewDevice a_device, viewHost a_host)
+  {
+    Kokkos::parallel_for("fill a_device", num_elements, KOKKOS_LAMBDA(int i)
```
Comment (dalg24) [inline core/unit_test/incremental_testing/TestIncremental_02_MemorySpace.hpp]: I don't think you are allowed to use `parallel_for` yet.

File: core/unit_test/incremental_testing/TestIncremental_02_MemorySpace.hpp (hunk 11 above)
Comment (dalg24) [reply core/unit_test/incremental_testing/TestIncremental_02_MemorySpace.hpp]: Please ignore moving to #2537

File: core/unit_test/incremental_testing/TestIncremental_02_MemorySpace.hpp [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+    Kokkos::parallel_reduce("compare_equal", num_elements, KOKKOS_LAMBDA(const int i, int& errorUpdate)
+    {
+      if(a_device(i) != 0.5) errorUpdate++;
+    }, error);
+  }
+
+  int compare_equal_host(viewHost a_host)
+  {
+    int error = 0;
+    Kokkos::parallel_reduce("compare_equal", num_elements, KOKKOS_LAMBDA(const int i, int& errorUpdate)
```
Comment (Rombur) [inline core/unit_test/incremental_testing/TestIncremental_02_MemorySpace.hpp]: `parallel_reduce` is a pretty advanced feature to be used in such an early test

File: core/unit_test/incremental_testing/TestIncremental_03_ParallelFor.hpp [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+/*
+ * Unit Test for Kokkos::parallel_for and Kokkos::parallel_reduce
```
Comment (Rombur) [inline core/unit_test/incremental_testing/TestIncremental_03_ParallelFor.hpp]: I think that we should split the test in two: one for `parallel_for` and one for `parallel_reduce`.

File: core/unit_test/incremental_testing/Test02a_MemorySpace.hpp [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+template <class MemSpace>
+struct TestIncrMemorySpace_malloc {
+
+  const int num_elements = 10;
+
+  void testit_malloc() {
+    int *data = (int*) Kokkos::kokkos_malloc<MemSpace>("data", num_elements*sizeof(int));
+    ASSERT_FALSE(data == nullptr);
+    Kokkos::kokkos_free<MemSpace>(data);
+  }
```
Comment (crtrott) [inline core/unit_test/incremental_testing/Test02a_MemorySpace.hpp]: This is a good first test.

Comment (crtrott) [reply core/unit_test/incremental_testing/Test02a_MemorySpace.hpp]: Lets add another one test (testit_multi_malloc or so), allocate multiple pointers, and make sure that the std::distance between all of them is larger than num_elements.

File: core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp (hunk 9 above)
Comment (crtrott) [reply core/unit_test/incremental_testing/Test01_ExecutionSpace.hpp]: You menthe second one?

**Comment by rgayatri23** [comment]
All the requested changes have been addressed.

**Comment by dalg24** [comment]
Retest this please

File: core/unit_test/incremental_testing/Test02a_MemorySpace.hpp (hunk 15 above)
Comment (janciesko) [reply core/unit_test/incremental_testing/Test02a_MemorySpace.hpp]: Ok.

**Pull request merged**